### PR TITLE
#197 relational.write: UPSERT statements should be generated by SQLAlchemy's SQL builder 

### LINKED ISSUE #197: `relational.write`: UPSERT statements should be generated by SQLAlchemy's SQL builder
there is an sql builder for the upsert now in SA. https://docs.sqlalchemy.org/en/14/dialects/postgresql.html#insert-on-conflict-upsert.
we should move to the API in next iteration as we add more dialects.

### PR DESCRIPTION
**Pull request by spicy-sauce**

### CODE REVIEW

File: core/src/datayoga_core/blocks/relational/write/block.py [hunk 1]
```diff
@@ -89,14 +91,13 @@ def init(self, context: Optional[Context] = None):
...
-            primary_keys = get_fields(self.keys)
-            mapping_fields = get_fields(self.mapping)
+            primary_keys = get_column_mapping(self.keys)
 
             self.delete_stmt = self.tbl.delete(
                 sa.and_(*[sa.text(f"{self.tbl.columns.get(field['column'])} = {sa.bindparam(field['key'])}")
                           for field in primary_keys]))
 
-            self.upsert_stmt = generate_upsert_stmt(self.tbl.fullname, primary_keys, mapping_fields, db_type)
+            self.upsert_stmt = generate_upsert_stmt(self.tbl, primary_keys, db_type)
```
Comment (zalmane) [inline core/src/datayoga_core/blocks/relational/write/block.py]: technically these aren't necessarily Primary keys. they are business keys

File: core/src/datayoga_core/blocks/relational/write/block.py [hunk 2]
```diff
@@ -134,15 +135,17 @@ def execute_upsert(self, records: List[Dict[str, Any]]):
             try:
                 get_key_values(record, self.keys)
             except ValueError as e:
-                record[Block.RESULT_FIELD] = Result(Status.REJECTED, f"{e}")
+                record[Block.RESULT_FIELD] = result.reject(f"{e}")
```
Comment (zalmane) [inline core/src/datayoga_core/blocks/relational/write/block.py]: didn't we change this already?

File: core/src/datayoga_core/blocks/relational/write/block.py [hunk 3]
```diff
@@ -19,40 +21,40 @@ class OpCode(Enum):
...
+                raise ValueError(f"{field['column']} column does not exist in {table.fullname} table")
 
-        insert_fields = ", ".join([field["column"] for field in mapping_fields])
-        pk_fields = ", ".join([field["column"] for field in primary_keys])
+        insert_stmt = insert(table)
 
-        insert_bind_params = ", ".join([f"{sa.bindparam(field['key'])}" for field in mapping_fields])
+        do_update_stmt = insert_stmt .on_conflict_do_update(
+            index_elements=[table.columns[field["column"]] for field in primary_keys],
+            set_=insert_stmt.excluded)
```
Comment (zalmane) [inline core/src/datayoga_core/blocks/relational/write/block.py:47]: why is the variable "set_"?

File: core/src/datayoga_core/blocks/relational/write/block.py (hunk 2 above)
Comment (spicy-sauce) [reply core/src/datayoga_core/blocks/relational/write/block.py]: yes, reverted.

File: core/src/datayoga_core/blocks/relational/write/block.py (hunk 3 above)
Comment (spicy-sauce) [reply core/src/datayoga_core/blocks/relational/write/block.py:47]: See here:
https://docs.sqlalchemy.org/en/14/dialects/postgresql.html#insert-on-conflict-upsert

This is the set clause in the `DO UPDATE SET`

File: core/src/datayoga_core/blocks/relational/write/block.py (hunk 1 above)
Comment (spicy-sauce) [reply core/src/datayoga_core/blocks/relational/write/block.py]: renamed

**Review by zalmane — APPROVED** [review]

**Pull request merged**